### ISSUE TITLE
pDNN score saturation is breaking category-boundary derivation grid-wide

### ISSUE THREAD

**Issue by raj2022** [raj2022/hhbbgg_AwkwardAnalyzer#129]
<html><head></head><body><h1>pDNN score saturation is breaking category-boundary derivation grid-wide</h1>
<h2>Summary</h2>
<p>The α(score) category-boundary search (<code>build_edges()</code>, per B2G-24-001's
greedy AMS² method) produces degenerate, non-informative boundaries --
effectively <code>[1.0, 1.0, 1.0, 1.0, 1.0]</code> -- for every mass point checked
so far. Root cause confirmed directly: a large, consistent fraction of
signal-region signal MC events have <code>pDNN_score</code> saturated at (or
extremely close to) the ceiling value of 1.0, well beyond the <code>--nmin</code>
event count the greedy search needs to define its first candidate
category.</p>
<h2>Evidence</h2>
<p>Checked four mass points spanning the grid, <code>selection</code>-region SR events
with <code>pDNN_score &gt;= 0.9999</code>:</p>

Mass point | SR events | At score ceiling | Fraction
-- | -- | -- | --
X=300,  Y=90 | 14,472 | 1,234 | 8.5%
X=600,  Y=400 | 17,895 | 2,312 | 12.9%
X=950,  Y=700 | 20,456 | 3,133 | 15.3%
X=1000, Y=100 | 29,317 | 16,612 | 56.7%

<p>In every single case, the score of the 50th-highest-ranked SR event
(matching <code>--nmin 50</code>, the default candidate-bin size) is exactly
<code>1.000000</code> -- meaning the greedy search's very first accepted category
boundary is carved entirely out of the saturated cluster before ever
reaching a score value that reflects real separation power.</p>
<p>Confirmed this is not a region-selection artifact: the same saturation
pattern (and, for the two regions checked, a <em>worse</em> saturation
fraction in the tighter <code>srbbgg</code> region than in <code>selection</code>) holds
across every region tested. Switching regions does not resolve this.</p>
<p>Confirmed this is not a background-statistics artifact either:
background MC and data both have ample SR/CR statistics with a healthy,
non-degenerate <code>pDNN_score</code> spread (mean ~0.10-0.20, full [0,1] range) --
the problem is specifically the signal side's score distribution.</p>
<h2>Why this breaks the categorization method specifically</h2>
<p><code>AMS(s, b)</code> is defined to return exactly 0 whenever <code>b &lt;= 0</code>, and the
greedy search's gain formula treats a <code>best</code> significance of 0 as an
automatic accept (<code>gain = 1.0</code>) for the next candidate bin. Combined
with tens of thousands of signal events sharing the identical score
value, every one of the <code>--max-bins</code> accepted categories ends up drawn
from the same saturated pocket -- the algorithm never has a reason to
look past it, regardless of <code>--nmin</code>, region choice, or ttH-killer
pre-cut.</p>
<h2>Open question -- not yet resolved, needs input</h2>
<p>Is this saturation:</p>
<ol>
<li><strong>Genuine, expected separation</strong> -- for at least some mass points, a
well-trained network being extremely confident on most on-shell
signal MC, or</li>
<li><strong>Miscalibration / overconfidence</strong> -- a common failure mode,
especially plausible if evaluated on the same or similar MC used in
training, or</li>
<li>Something else in how the score is computed/stored (e.g. extreme
logits clipped numerically to exactly 1.0 by <code>sigmoid()</code>).</li>
</ol>
<p>The clear mass-dependence (X=1000/Y=100 saturating 4-6x more heavily
than the other three points checked) is itself worth understanding --
could reflect genuinely easier kinematic separation at that corner of
the grid, or a training-data-density effect (fewer/more examples near
that point during pDNN training), or something else not yet identified.</p>
<h2>What this is NOT</h2>
<ul>
<li>Not a bug in the categorization script's region logic (<code>selection</code>
vs <code>srbbgg</code> -- both saturate).</li>
<li>Not a background-statistics or <code>α(score)</code> reweighting problem --
background/data inputs are healthy.</li>
<li>Not specific to one mass point -- confirmed at four points spanning
the grid, all showing the same qualitative failure.</li>
</ul>
<h2>Next steps (pending decision)</h2>
<ul>
<li>[ ] Confirm with whoever owns the pDNN training whether this level of
score saturation is expected/known, or a symptom worth
investigating in training/calibration.</li>
<li>[ ] If genuine and expected: <code>build_edges()</code> itself needs a fix
independent of any region/nmin tuning -- e.g. deduplicating on
distinct score values before counting toward <code>--nmin</code>, so a
single candidate bin can't be drawn entirely from one saturated
spike.</li>
<li>[ ] If a miscalibration/training issue: may need revisiting at the
pDNN training level before categorization can proceed
meaningfully, since every downstream category boundary depends on
score separation actually reflecting confidence, not saturation.</li>
<li>[ ] Either way: do not treat the current <code>event_categories.json</code>
(from the run that produced this finding) as usable -- it reflects
the degenerate boundaries described above.</li>
</ul></body></html>
